proposal: react component architecture (#66)
eigen-space/codestyle: doc/architecture/frontend/README.md

# Architecture

## 1. Component

In our projects we follow this architecture:
![Architecture](assets/architecture.png)

> ### 1.1 Smart group

Reviewer: ```suggestion
### 1.1. Component Logic layer
```

Author: done

> Logic and data processing. Usual supported by developers.

Reviewer: ```suggestion
Logic and data processing. Usually maintained by developers.
```

Author: done

#### 1.1.1 Container

Main question:\

> What does business data component show and business logic

Reviewer: ```suggestion
What business data does component operate with? 
What business logic does component implement?
```

Author: done

component use? 

Responsible for:
1. Getting and preparing data from store
2. Preparing actions

> 3. Including business logic

Reviewer: ```suggestion
3. Implementation of business logic
```

Author: done

Filename:\
`<component>.container.tsx`

> Exports `mapStateToProps` and / or `mapDispatchToProps` functions.

Reviewer: Do we want to export functions? Maybe should we export connected component? We export functions to be able to test them, right? However if we do not export connected component, there is something wrong in naming of this file because container is a connected component.

Example:
```
export const mapStateToProps = (state: StoreState): MappingStateProps => ({
    ownCards: getOwnCards(state),
    cardsForList: getCardsForList(state)
});
```

#### 1.1.2 Presenter

Stateful component.

Main question:\
What does component do?

Responsible for:
1. Storing and controlling UI state

> 2. Including all methods

Reviewer: ```suggestion
2. Implementation of all methods like handlers and component logic
```

Author: done

Filename:\
`<component>.presenter.tsx`

Exports class name:\
`<Component>Presenter`

Example:
```
export class DomainListPresenter extends React.PureComponent<Props> {
    
    ...

    render(): React.ReactNode {
        const { ownCards, cardsForList } = this.props;
    
        return (
            <DomainListView
                items={ownCards}
                cardsForList={cardsForList}
                onCardUpdate={this.onCardUpdated}
                onCardAction={this.onCardAction}
            />
        );
    }

    private onDomainCreated(id: Guid): void {
        this.props.history.push(`${RouteType.DOMAINS}/${id}`);
    }

    private onCardUpdated(entity: CardEntity): void {
        const domainArgs = this.cardAppService.convertCardToDomainArgs(entity);
        this.domainActions.update(domainArgs);
    }

    ...
}
```

> ### 1.2 Stub group

Reviewer: ```suggestion
### 1.2 Presentation Layer
```

Author: done

Only UI view layers. Can be supported by designers.

#### 1.2.1 Styles 

Main question:\
How does component look?

Filename:\
`<component>.styles.tsx`

Example:
```
...

export const DomainListRoot = styled.div`
    position: relative;
    background: ${colors.background};
    height: 100%;
`;

export const CardListWrapper = styled(ContainerRoot)`
    height: ${styles.searchSectionPadding}%;
    overflow-y: auto;
`;

...
```

#### 1.2.2 View 

Main question:\
What does component show?

Responsible for:
1. Stateless component

> 2. Contains only makeup with handlers and interpolate data

Reviewer: ```suggestion
2. Only makeup with handler interfaces and data interpolation
```

Author: done

 
Filename:\
`<component>.view.tsx`

Exports class name / functional component name:\
`<Component>View`

Example:
```
interface Props {
    items: CardEntity[];
    cardsForList: CardEntity[];
    onCardUpdate: (entity: CardEntity) => void;
    onCardAction: (action: CardContextMenuAction, entity: CardEntity) => void;
}

> function DomainList({ items, cardsForList, onCardUpdate, onCardAction }: Props): React.ReactElement {

Reviewer: ```suggestion
function DomainListView({ items, cardsForList, onCardUpdate, onCardAction }: Props): React.ReactElement {
```

Author: done

    return (
        <DomainListRoot>
            <CardListWrapper>
                <CardList
                    items={items}
                    onCardEdit={onCardUpdate}
                    onAction={onCardAction}
                />
            </CardListWrapper>
        </DomainListRoot>
    );
}

export const DomainListView = React.memo(DomainList);
```

> #### 1.2.3 Animation

Reviewer: ```suggestion
#### 1.2.3. Animation
```

Author: done

Main question:\
How does component change UI states?

Responsible for:
1. Includes animation states for some components.
2. As Framer Motion, as CSS

Filename:\
`<component>.animation.ts`

Example:
```
import { SwitchState } from '../switch.enum.ts';

const transition = { type: 'tween', duration: 1 };

export const animationStates = {
    container: {
        [SwitchState.OFF]: {
            background: '#bbbbbb',
            transition
        },
        [SwitchState.ON]: {
            background: '#0070df',
            transition
        }
    },
};
```

> ## 1.3 Index

Reviewer: ```suggestion
## 1.3. Index
```

Author: done
